### DISCUSSION TITLE
Undo button

### DISCUSSION THREAD

**Discussion by kkhanmd** [pybricks/support#1878] [Ideas]
As we code few things we need are
1) Undo button, while trying to use touch tablet we end up dragging some code inadvertently. It would be great to have an undo button 
2) see battery percentage 
3) ability to download and run multiple programs from the brick menu
4) Copy a stack of block codes, currently only option is copy each individual block code or copy all, there is no option to multiselect or select all the blocks in a stack
5) able to place a program in view only mode, so nothing gets moved. Just a simple lock or unlock button should suffice. It would be better if you could copy in view mode.

**Comment by dlech** [comment]
> 3\. ability to download and run multiple programs from the brick menu

Can you please describe this one in more detail?

**Reply by kkhanmd** [reply]
Could you help me direct how to make a hub menu using block codes? My team has multiple missions coded using block codes and dont know how to store them in the hub for the competition. I saw some documentation using text based coding, but kids do not know how to use text base coding. https://pybricks.com/project/spike-hub-menu/
Is it possible to download multiple missions on to the hub using block coding?

**Reply by laurensvalk** [reply]
All of your programs can be block programs. The Python program is just for the menu.

You can copy that example. All you have to do is change the program names in your menu.

There will also be a default menu in the brick at some point, but this may be a while.

**Reply by laurensvalk** [reply]
If you can make a list of your program names, I can show you what the menu would be :slightly_smiling_face:

**Comment by kkhanmd** [comment]
![Screenshot_20241021_113554_Chrome](https://github.com/user-attachments/assets/90218556-c366-48b6-a16a-ebdd39425eca)
I tried running this, what's the syntax error in line 12?

**Reply by laurensvalk** [reply]
It looks like copy-pasting added several extra spaces in your code. You'll want to delete those extra ones. Have a look at this:

![image](https://github.com/user-attachments/assets/6fd77a0e-0e1b-4f58-b9a7-5dbc827d93c9)

**Comment by kkhanmd** [comment]
![Screenshot_20241021_121036_Chrome](https://github.com/user-attachments/assets/b3e1b89b-9c67-464b-9751-a412db5d9f6b)
Ok the program is running, but I don't see anything on the hub, I was expecting to see 1, 2  or 3 but the hub just shows a block of 3x3 yellow lights unchanged from what I started with, when I press the forward and backward button, nothing happens

**Reply by BertLindeman** [reply]
Hello,


Maybe a long shot, but did you start the program by pressing the hub button or in the pyrbicks user interface by pressing the run button or F5?
Explanation: If you update a program in pybricks code or in beta.pybricks.com the program is "compiled" and loaded into the hub.

If you press the hub button, the latest loaded program in the hub is started. This does not update the program in the hub.

If you update the hub menu program or any of the missions, that requires you to load the programs into the hub again. 
Running the main hub (menu) program from code.pybricks.com is enough. The missions get loaded in that case as well.

**Reply by BertLindeman** [reply]
The programs I tested with:

[EDIT] remove the while loop in the main program
`menu_example.py`
```python

from pybricks.tools import hub_menu

selected = hub_menu("1", "2", "3")

# Based on the selection, run a program.
if selected == "1":
    import mission1
elif selected == "2":
    import mission2
elif selected == "3":
    import mission3

```

`mission1.py`
```python

from pybricks.hubs import PrimeHub
from pybricks.tools import wait

HUB = PrimeHub()
HUB.display.char("A")
wait(4000)

```

`mission2.py`
```python

from pybricks.hubs import PrimeHub
from pybricks.tools import wait

HUB = PrimeHub()
HUB.display.char("B")
wait(4000)

```

`mission3.py`
```python

from pybricks.hubs import PrimeHub
from pybricks.tools import wait

HUB = PrimeHub()
HUB.display.char("C")
wait(4000)

```

**Reply by laurensvalk** [reply]
@BertLindeman, you cannot import things in a loop. Just the one-off example is probably easier for beginners.

**Reply by laurensvalk** [reply]
Are there any errors at the bottom? The slider is dragged all the way down so it is a bit hard to see.

**Reply by kkhanmd** [reply]
Thanks for the reply, the error says Primehub isnt defined. Our Robot is named Mercury, is that the reason? where does that go?

**Reply by BertLindeman** [reply]
> @BertLindeman, you cannot import things in a loop. Just the one-off example is probably easier for beginners.

Removed the loop from the main program.

**Comment by kkhanmd** [comment]
![Screenshot_20241021_154309_Chrome](https://github.com/user-attachments/assets/417eca77-8454-47cd-9e4b-f4101b8395fe)

**Reply by dlech** [reply]
Should be:

```python
from pybricks.hubs import PrimeHub

hub = PrimeHub()
```

instead of:

```python
hub = Mercury()
```

**Comment by kkhanmd** [comment]
Ok it's fixed, I deleted the hub menu file and started from scratch. This time I unselected use template before making the file and everything works. 
![Screenshot_20241021_155848_Chrome](https://github.com/user-attachments/assets/d3e7e5c9-0cae-44df-862a-8f61b7ae4d1b)

**Comment by kkhanmd** [comment]
![Screenshot_20241026_214356_Chrome](https://github.com/user-attachments/assets/673ca0ef-1841-463a-8a7f-ace93d38f0b8)
I am still having issues with my Hub menu, can you please check and advice?  In this instance, I can download Mission 1, 2 and 3 and can toggle between the three missions, but once I run any of the missions, the hub menu disappears. I thought the purpose of hub menu was to run the second program after the first one finishes. Let me know what I am doing wrong.

**Reply by laurensvalk** [reply]
The menu is just like any other program.

So you can just press the main button on the hub again to open the menu again.

**Comment by BertLindeman** [comment]
Hello,

1. After a mission has been done, the main program stops.
When you are ready to start  another mission: Press the button on the hub to start the main program again, and choose that misson.

2. Or follow the explanation of the Caveats section in [[How to download multiple pybricks programs with your own menu](https://pybricks.com/project/spike-hub-menu/)](https://pybricks.com/project/spike-hub-menu/#multiple-mission-programs) :

Part of the explanation:
```
The approach used above works great for a basic menu, 
but you might want to organize your code a bit differently
in case you want to take this menu technique even further.

The import statement will run a program only the first time,
even if you run the same import again in the same program.

This is why programmers would usually write the “mission” as a function (a task) instead.
You could import that function and call it from your menu program as many times as you like.
```

I hope this helps.

[EDIT] a bit better formatting
Bert

**Comment by kkhanmd** [comment]
Awesome, that works, one more question, Is there a way to  write a hub menu so that it automatically advance to the next Mission? so for example I run Mission 1, now it auto advance to mission 2 and I have to only hit the center button once to start it?

**Reply by BertLindeman** [reply]
We are a long way drifted from the original post of this discussion, but anyway:

If you automatically want to run a number of missions after each other, would you need a menu then?

Or do you want to select in a menu at which mission to start with?

That might be something like:
```python
from pybricks.tools import hub_menu

# Make a menu to choose a letter or number.
selected = hub_menu("1", "2", "3")

# Based on the selection, run the missions.
if selected == "1":
    import mission1
    import mission2
    import mission3
elif selected == "2":
    import mission2
    import mission3
elif selected == "3":
    import mission3

```

**Reply by kkhanmd** [reply]
Thanks for answering this, what I really want to do is when ever a mission is completed, the brick is showing me the next mission number, waiting for me to press the center button to start the mission. In the above example, I run the hub menu, and then select 1 for mission 1. Once mission 1 is complete what I want is for  my brick to  show 2 on it, so I can just press the center button to run mission 2 when ready, instead of pressing the center button to run the hub menu and then pressing the forward button to scroll to 2. Is there a way to write the hub menu, so after it finishes one mission, it is automatically displaying the next mission to run with a single press of the center button? My kids may have 9 missions to run and if they are on number 5 for example, it would take a long time and multiple clicks to reach mission 6 to start.  Thank you for all the help.

**Reply by BertLindeman** [reply]
Hi,

My example on [the previous post](https://github.com/orgs/pybricks/discussions/1878#discussioncomment-11068997)  does just that, you just want to get 6 more missions in it to get the 9 missions complete.

You might want to add something between the missions to wait for a button-press to start the next mission.

I think you can do that.

Bert

**Comment by kkhanmd** [comment]
Ok I have zero coding experience, that's why I bought the block version. Is this how the menu should look?(I'm sorry I do healthcare and trying to lead kids on this)
![Screenshot_20241101_101301_Chrome](https://github.com/user-attachments/assets/05a3d51a-c24d-4d63-9149-700d2999d7df)

**Reply by BertLindeman** [reply]
Coming close.
A good first time python start.
This could also be done in block coding.
But you are trying python, so let us continue in python for now.

We are going to use more things, so we need some more things to import:
```python
from pybricks.hubs import PrimeHub
from pybricks.tools import hub_menu
from pybricks.tools import wait
from pybricks.parameters import Button
```

After each mission you want to wait until the center button is pressed again.
And because of the speed of the hub, you will want to wait until the center button is released again.

This is done without testing, so no guarantee.

To ease maintenance on the program I define the "next" button as a variable we can use:
```python
my_next_button = Button.RIGHT
```

I choose the right button, as the center button would **stop** the program without further action, 
so the right button to start the next mission is a safe one.

The start of the program might look like this:
```python

from pybricks.hubs import PrimeHub
from pybricks.tools import hub_menu
from pybricks.tools import wait
from pybricks.parameters import Button

hub = PrimeHub()

# Make a menu to choose a letter or number.
selected = hub_menu("1", "2", "3")
my_next_button = Button.RIGHT
# Based on the selection, run the missions.
if selected == "1":
    import mission1
    hub.display.char(">")
    while not my_next_button in hub.buttons.pressed():
        wait(10)
    while my_next_button in hub.buttons.pressed():
        wait(10)
    import mission2
```

Would you get further with these hints?

Have fun. Python **is** fun,

Bert

**Comment by kkhanmd** [comment]
![Screenshot_20241103_214237_Chrome](https://github.com/user-attachments/assets/2f20702f-920c-4925-9400-e9d005ef5251)
So I ran that last night and nothing changed. After my mission 1 ran, it only showed either the ">" or "2" depending upon what I put in there, but pressing the right, center or left button did nothing. There was no automatic change of the display to show Mission 2 is ready to be run with a single press of a button.

**Reply by BertLindeman** [reply]
Sorry if I gave you the impression that I presented a complete program.
The idea is that you make a complete program with it.

To test I use very simple mission programs to prevent confusion.
So you can see how to menu program acts. 
Each mission program shows another letter and beeps a different tone.
Something like this:
```python
from pybricks.hubs import PrimeHub
from pybricks.tools import wait

hub = PrimeHub()
hub.display.char("A")
hub.speaker.beep(440, 50)
wait(1000)
```

Remember that pressing the center button normally stops the running program.
If you want to prevent the center button to end the program, add this to the program:
```python
# Press of the center button now does NOT stop the program 
hub.system.set_stop_button(None)
```

To help you further a complete program that handles two missions.
Phase 1:
 - use left and right button to select the mission.
 - press the center button to start the selected mission.
 - 
Phase 2:
- press the right button to start the next mission.

```python

from pybricks.hubs import PrimeHub
from pybricks.tools import hub_menu
from pybricks.tools import wait
from pybricks.parameters import Button

hub = PrimeHub()
# Press of the center button now does NOT stop the program 
hub.system.set_stop_button(None)

# Make a menu to choose a letter or number.
selected = hub_menu("1", "2")
my_next_button = Button.RIGHT
# Based on the selection, run the missions.
if selected == "1":
    import mission1
    print("mission 1 ended")
    hub.display.char(">")
    while not my_next_button in hub.buttons.pressed():
        wait(10)
    while my_next_button in hub.buttons.pressed():
        wait(10)
    import mission2
    print("mission 2 ended")
elif selected == "2":
    import mission2
    print("mission 2 ended")
print("missions 1 and 2 ended")
```

Bert

**Comment by zachflanders** [comment]
Hi @kkhanmd 

Try this out:
```python
import importlib

from pybricks.hubs import PrimeHub
from pybricks.parameters import Button
from pybricks.tools import wait

program_running = True

# list of the mission modules to run
missions = [
    "mission1",
    "mission2",
    "mission3",
    "mission4",
    "mission5",
    "mission6",
    "mission7",
    "mission8",
    "mission9",
    "mission10",
]

hub = PrimeHub()

# Press of the center button now does NOT stop the program
hub.system.set_stop_button(None)

# set the control button to the center button
control_button = Button.CENTER

while program_running:
    for mission in missions:
        print(f"Press the center button to start {mission}.")
        hub.display.char("R")  # Display "R" for "Ready"
        # Wait for the control button to be pressed
        while not control_button in hub.buttons.pressed():
            wait(10)
        print(f"Starting {mission}.")
        # Display the mission number (last character of the mission name)
        hub.display.char(mission[-1])
        importlib.import_module(mission)
        print(f"Finished {mission}.")
    program_running = False
print("All missions completed.")
```
When the code is run it should do the following:
 - Print `Press the center button to start mission1.` to the console
 - The hub should show `R` for "Ready"
 - It will then wait and do nothing until you press the center button
 - Once you press the center button it will print `Starting mission1.` and the hub will display `1`
 - It will import the mission1 module, running the mission1 code in the process
 - It will then print `Finished mission1.` in the console.
 - It will immediately print `Press the center button to start mission2.`,the hub will switch to `R`, and it will wait until you press the center button
 - ...etc until it completes the last mission at which point it will print `All missions completed.`

**Reply by laurensvalk** [reply]
Thank you. `importlib` is not available on SPIKE Prime. It uses MicroPython, which is a small implementation of Python that does not have all of its features.

**Reply by zachflanders** [reply]
I have not tested, but I believe you could replace with `__import__()`
```python
from pybricks.hubs import PrimeHub
from pybricks.parameters import Button
from pybricks.tools import wait

program_running = True

# list of the mission modules to run
missions = [
    "mission1",
    "mission2",
    "mission3",
    "mission4",
    "mission5",
    "mission6",
    "mission7",
    "mission8",
    "mission9",
    "mission10",
]

hub = PrimeHub()

# Press of the center button now does NOT stop the program
hub.system.set_stop_button(None)

# set the control button to the center button
control_button = Button.CENTER

while program_running:
    for mission in missions:
        print(f"Press the center button to start {mission}.")
        hub.display.char("R")  # Display "R" for "Ready"
        # Wait for the control button to be pressed
        while not control_button in hub.buttons.pressed():
            wait(10)
        print(f"Starting {mission}.")
        # Display the mission number (last character of the mission name)
        hub.display.char(mission[-1])
        __import__(mission)
        print(f"Finished {mission}.")
    program_running = False
print("All missions completed.")
```

**Comment by laurensvalk** [comment]
> Copy a stack of block codes, currently only option is copy each individual block code or copy all, there is no option to multiselect or select all the blocks in a stack

You can hold `shift` and then drag a selection. Or hold `shift` and click on the blocks you want.
